File: libsignal/service/src/main/java/org/whispersystems/signalservice/api/kbs/MasterKey.java
package org.whispersystems.signalservice.api.kbs;

import org.whispersystems.signalservice.api.storage.StorageKey;
import org.whispersystems.signalservice.internal.util.Hex;
import org.whispersystems.util.StringUtil;

import java.security.SecureRandom;
import java.util.Arrays;

import static org.whispersystems.signalservice.api.crypto.CryptoUtil.hmacSha256;

public final class MasterKey {

  private static final int LENGTH = 32;

  private final byte[] masterKey;

  public MasterKey(byte[] masterKey) {
    if (masterKey.length != LENGTH) throw new AssertionError();

    this.masterKey = masterKey;
  }

  public static MasterKey createNew(SecureRandom secureRandom) {
    byte[] key = new byte[LENGTH];
    secureRandom.nextBytes(key);
    return new MasterKey(key);
  }

  public String deriveRegistrationLock() {
    return Hex.toStringCondensed(derive("Registration Lock"));
  }

  public StorageKey deriveStorageServiceKey() {
    return new StorageKey(derive("Storage Service Encryption"));
  }

  private byte[] derive(String keyName) {
    return hmacSha256(masterKey, StringUtil.utf8(keyName));
  }

  public byte[] serialize() {
    return masterKey.clone();
  }

  @Override
  public boolean equals(Object o) {
    if (o == null || o.getClass() != getClass()) return false;

    return Arrays.equals(((MasterKey) o).masterKey, masterKey);
  }

  @Override
  public int hashCode() {
    return Arrays.hashCode(masterKey);
  }
}

File: core/models-jvm/src/main/java/org/signal/core/models/MasterKey.kt
/*
 * Copyright 2025 Signal Messenger, LLC
 * SPDX-License-Identifier: AGPL-3.0-only
 */
package org.signal.core.models

import org.signal.core.models.storageservice.StorageKey
import org.signal.core.util.Base64
import org.signal.core.util.CryptoUtil
import org.signal.core.util.Hex
import java.security.SecureRandom

class MasterKey(masterKey: ByteArray) {
  private val masterKey: ByteArray

  companion object {
    private const val LENGTH = 32

    fun createNew(secureRandom: SecureRandom): MasterKey {
      val key = ByteArray(LENGTH)
      secureRandom.nextBytes(key)
      return MasterKey(key)
    }
  }

  init {
    check(masterKey.size == LENGTH) { "Master key must be $LENGTH bytes long (actualSize: ${masterKey.size})" }
    this.masterKey = masterKey
  }

  fun deriveRegistrationLock(): String {
    return Hex.toStringCondensed(derive("Registration Lock"))
  }

  fun deriveRegistrationRecoveryPassword(): String {
    return Base64.encodeWithPadding(derive("Registration Recovery")!!)
  }

  fun deriveStorageServiceKey(): StorageKey {
    return StorageKey(derive("Storage Service Encryption")!!)
  }

  fun deriveLoggingKey(): ByteArray? {
    return derive("Logging Key")
  }

  private fun derive(keyName: String): ByteArray? {
    return CryptoUtil.hmacSha256(masterKey, keyName.toByteArray(Charsets.UTF_8))
  }

  fun serialize(): ByteArray {
    return masterKey.clone()
  }

  override fun equals(o: Any?): Boolean {
    if (o == null || o.javaClass != javaClass) return false

    return (o as MasterKey).masterKey.contentEquals(masterKey)
  }

  override fun hashCode(): Int {
    return masterKey.contentHashCode()
  }

  override fun toString(): String {
    return "MasterKey(xxx)"
  }
}
